opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -12,7 +12,214 @@
+        self.maxlen = maxlen
+        self.dropped = 0
+        self._dict = OrderedDict()
+        self._lock = threading.Lock()
+
+    def __repr__(self):
+        return ("{}({}, maxlen={})"
+                .format(
+                    type(self).__name__,
+                    dict(self._dict),
+                    self.maxlen
+                ))
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __setitem__(self, key, value):
+        with self._lock:
+            if key in self._dict:
+                del self._dict[key]
+            elif len(self._dict) == self.maxlen:
+                del self._dict[next(iter(self._dict.keys()))]
+                self.dropped += 1
+            self._dict[key] = value
+
+    def __delitem__(self, key):
+        del self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    @classmethod
+    def from_map(cls, maxlen, mapping):
+        mapping = OrderedDict(mapping)
+        if len(mapping) > maxlen:
+            raise ValueError
+        bounded_dict = cls(maxlen)
+        bounded_dict._dict = mapping
+        return bounded_dict
+
+
+class SpanContext(trace_api.SpanContext):
+    """See `opentelemetry.trace.SpanContext`."""
+
+    def is_valid(self) -> bool:
+        return (self.span_id == trace_api.INVALID_SPAN_ID or
+                self.trace_id == trace_api.INVALID_SPAN_ID)
+
+
+Event = namedtuple('Event', ('name', 'attributes'))
+
+Link = namedtuple('Link', ('context', 'attributes'))
+
+
+class Span(trace_api.Span):
+    def __init__(self: 'Span',
+                 name: str,
+                 context: 'SpanContext',
+                 # TODO: span processor
+                 parent: typing.Union['Span', 'SpanContext'] = None,
+                 root: bool = False,
+                 sampler=None,  # TODO
+                 trace_config=None,  # TraceConfig TODO
+                 resource=None,  # Resource TODO
+                 # TODO: is_recording
+                 attributes=None,  # type TODO
+                 events=None,  # type TODO
+                 links=None,  # type TODO
+                 ) -> None:
+        """See `opentelemetry.trace.Span`."""
+        if root:
+            if parent is not None:
+                raise ValueError("Root span can't have a parent")
+
+        self.name = name
+        self.context = context
+        self.parent = parent
+        self.root = root
+        self.sampler = sampler
+        self.trace_config = trace_config
+        self.resource = resource
+        self.attributes = attributes
+        self.events = events
+        self.links = links
+
+        if attributes is None:
+            self.attributes = BoundedDict(MAX_NUM_ATTRIBUTES)
+        else:
+            self.attributes = BoundedDict.from_map(
+                MAX_NUM_ATTRIBUTES, attributes)
+
+        if events is None:
+            self.events = BoundedList(MAX_NUM_EVENTS)
```

Comment:
We don't have concrete measurement today, but I suspect it could be pretty expensive if most spans have no attributes/events/links.
Consider create a class variable `Span.copy_on_write_bounded_list/dict` and assign it as the default value. Allocate the real object when there are actual write operation.

Reply:
Changed in 87cac92, although it's probably worth profiling before making too many optimizations like this.